red_october/models/red_october_file.py

```diff
@@ -0,0 +1,186 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016-2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+import logging
+
+from odoo import _, api, fields, models
+from odoo.http import request
+from odoo.exceptions import ValidationError
+
+
+_logger = logging.getLogger(__name__)
+
+
+class RedOctoberFile(models.Model):
+
+    _name = 'red.october.file'
+    _description = 'Red October File'
+    _inherits = {'ir.attachment': 'attachment_id'}
+
+    attachment_id = fields.Many2one(
+        string='Attachment',
+        comodel_name='ir.attachment',
+        required=True,
+        ondelete='cascade',
+    )
+    owner_ids = fields.One2many(
+        string='Owners',
+        comodel_name='red.october.file.owner',
+        inverse_name='file_id',
+        help='These are the delegations required for decryption.',
+    )
+    vault_id = fields.Many2one(
+        string='Vault',
+        comodel_name='red.october.vault',
+        required=True,
+        default=lambda s: s.env['red.october.vault'].get_current_vault(),
+    )
+    delegation_min = fields.Integer(
+        string='Minimum Delegations',
+        default=1,
+        required=True,
+        help='Minimum amount of delegations that are required from file '
+             'owners in order to allow encryption.',
+    )
+
+    _sql_constraints = [
+        ('attachment_id_unique', 'UNIQUE(attachment_id)',
+         'This attachment already has an associated Red October file.'),
+    ]
+
+    @api.multi
+    @api.constrains('owner_ids')
+    def _check_owner_ids(self):
+        """ It will not allow the same owner twice.
+
+        Raises:
+            ValidationError: If the owner is duplicated on the record.
+        """
+        for record in self:
+            if len(record.owner_ids) != len(set(record.owner_ids)):
+                raise ValidationError(_(
+                    'This user has already been delegated permissions for '
+                    'this file.',
+                ))
+
+    @api.model
+    def default_get(self, fields):
+        result = super(RedOctoberFile, self).default_get(fields)
+        result['type'] = self.env['ir.attachment'].RED_OCTOBER
+        return result
+
+    @api.multi
+    def write(self, vals):
+        """ It will not allow vault transfers.
+
+        Raises:
+            NotImplementedError: If a vault transfer is being attempted.
+        """
+        if vals.get('vault_id'):
+            changed = self.filtered(
+                lambda r: r.vault_id.id != vals['vault_id']
+            )
+            if len(changed):
```

Comment:
your don't need to test the length for a recordset: https://github.com/OCA/OCB/blob/10.0/odoo/models.py#L5062